http_call.go

```diff
@@ -24,6 +25,102 @@ import (
 	"sync"
 )
 
+type unaryHTTPCall struct {
+	httpClient       HTTPClient
+	onRequestSend    func(*http.Request)
+	validateResponse func(*http.Response) *Error
+
+	request  *http.Request
+	response *http.Response // nil until the request has been sent
+}
+
+func newUnaryHTTPCall(
+	ctx context.Context,
+	httpClient HTTPClient,
+	url *url.URL,
+	header http.Header,
+) *unaryHTTPCall {
+	// ensure we make a copy of the url before we pass along to the
+	// Request. This ensures if a transport out of our control wants
+	// to mutate the req.URL, we don't feel the effects of it.
+	url = cloneURL(url)
+
+	// This is mirroring what http.NewRequestContext did, but
+	// using an already parsed url.URL object, rather than a string
+	// and parsing it again. This is a bit funny with HTTP/1.1
+	// explicitly, but this is logic copied over from
+	// NewRequestContext and doesn't effect the actual version
+	// being transmitted.
+	request := (&http.Request{
+		Method:     http.MethodPost,
+		URL:        url,
+		Header:     header,
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		Body:       nil, // Set in Do.
+		Host:       url.Host,
+	}).WithContext(ctx)
```

Comment:
We should consolidate this into a helper function instead of having this snippet and associated comment in here 2x.